File: core/src/main/java/com/netflix/conductor/core/execution/tasks/SetVariable.java
package com.netflix.conductor.core.execution.tasks;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import com.netflix.conductor.core.config.ConductorProperties;
import com.netflix.conductor.core.dal.ExecutionDAOFacade;
import com.netflix.conductor.core.exception.NonTransientException;
import com.netflix.conductor.core.execution.WorkflowExecutor;
import com.netflix.conductor.model.TaskModel;
import com.netflix.conductor.model.WorkflowModel;

import com.fasterxml.jackson.databind.ObjectMapper;

import static com.netflix.conductor.common.metadata.tasks.TaskType.TASK_TYPE_SET_VARIABLE;

@Component(TASK_TYPE_SET_VARIABLE)
public class SetVariable extends WorkflowSystemTask {

    private static final Logger LOGGER = LoggerFactory.getLogger(SetVariable.class);

    private final ConductorProperties properties;
    private final ObjectMapper objectMapper;

    private final ExecutionDAOFacade executionDAOFacade;

    public SetVariable(
            ConductorProperties properties,
            ObjectMapper objectMapper,
            ExecutionDAOFacade executionDAOFacade) {
        super(TASK_TYPE_SET_VARIABLE);
        this.properties = properties;
        this.objectMapper = objectMapper;
        this.executionDAOFacade = executionDAOFacade;
    }

    private boolean validateVariablesSize(
            WorkflowModel workflow, TaskModel task, Map<String, Object> variables) {
        String workflowId = workflow.getWorkflowId();
        long maxThreshold = properties.getMaxWorkflowVariablesPayloadSizeThreshold().toKilobytes();

        try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream()) {
            this.objectMapper.writeValue(byteArrayOutputStream, variables);
            byte[] payloadBytes = byteArrayOutputStream.toByteArray();
            long payloadSize = payloadBytes.length;

            if (payloadSize > maxThreshold * 1024) {
                String errorMsg =
                        String.format(
                                "The variables payload size: %d of workflow: %s is greater than the permissible limit: %d kilobytes",
                                payloadSize, workflowId, maxThreshold);
                LOGGER.error(errorMsg);
                task.setReasonForIncompletion(errorMsg);
                return false;
            }
            return true;
        } catch (IOException e) {
            LOGGER.error(
                    "Unable to validate variables payload size of workflow: {}", workflowId, e);
            throw new NonTransientException(
                    "Unable to validate variables payload size of workflow: " + workflowId, e);
        }
    }

    @Override
    public boolean execute(WorkflowModel workflow, TaskModel task, WorkflowExecutor provider) {
        Map<String, Object> variables = workflow.getVariables();
        Map<String, Object> input = task.getInputData();
        String taskId = task.getTaskId();
        ArrayList<String> newKeys;
        Map<String, Object> previousValues;

        if (input != null && input.size() > 0) {
            newKeys = new ArrayList<>();
            previousValues = new HashMap<>();
            input.keySet()
                    .forEach(
                            key -> {
                                if (variables.containsKey(key)) {
                                    previousValues.put(key, variables.get(key));
                                } else {
                                    newKeys.add(key);
                                }
                                variables.put(key, input.get(key));
                                LOGGER.debug(
                                        "Task: {} setting value for variable: {}", taskId, key);
                            });
            if (!validateVariablesSize(workflow, task, variables)) {
                // restore previous variables
                previousValues
                        .keySet()
                        .forEach(
                                key -> {
                                    variables.put(key, previousValues.get(key));
                                });
                newKeys.forEach(variables::remove);
                task.setStatus(TaskModel.Status.FAILED_WITH_TERMINAL_ERROR);
                return true;
            }
        }

        task.setStatus(TaskModel.Status.COMPLETED);
        executionDAOFacade.updateWorkflow(workflow);
        return true;
    }
}

File: core/src/execution/tasks/SetVariable.ts
import { WorkflowSystemTask } from '../WorkflowSystemTask.js';
import type { WorkflowModel, TaskModel } from '../types.js';
import type { WorkflowExecutor } from '../WorkflowExecutor.js';
import { TaskType } from '@agentmesh/common';

export class SetVariable extends WorkflowSystemTask {
  private maxPayloadSizeKb: number;

  constructor(maxPayloadSizeKb = 5120) {
    super(TaskType.SET_VARIABLE);
    this.maxPayloadSizeKb = maxPayloadSizeKb;
  }

  override execute(
    workflow: WorkflowModel,
    task: TaskModel,
    provider: WorkflowExecutor,
  ): boolean {
    const variables = workflow.variables;
    const input = task.inputData;
    if (input != null && Object.keys(input).length > 0) {
      const previousValues = new Map<string, unknown>();
      const newKeys: string[] = [];

      for (const key of Object.keys(input)) {
        if (key in variables) {
          previousValues.set(key, variables[key]);
        } else {
          newKeys.push(key);
        }
        variables[key] = input[key];
      }

      if (!this.validateVariablesSize(workflow, task, variables)) {
        for (const [key, val] of previousValues) {
          variables[key] = val;
        }
        for (const key of newKeys) {
          delete variables[key];
        }
        task.status = 'FAILED_WITH_TERMINAL_ERROR';
        return true;
      }
    }

    task.status = 'COMPLETED';
    return true;
  }

  private validateVariablesSize(
    workflow: WorkflowModel,
    task: TaskModel,
    variables: Record<string, unknown>,
  ): boolean {
    const payloadStr = JSON.stringify(variables);
    const payloadSize = Buffer.byteLength(payloadStr, 'utf-8');
    const maxThresholdBytes = this.maxPayloadSizeKb * 1024;

    if (payloadSize > maxThresholdBytes) {
      const errorMsg = `The variables payload size: ${payloadSize} of workflow: ${workflow.workflowId} is greater than the permissible limit: ${this.maxPayloadSizeKb} kilobytes`;
      task.reasonForIncompletion = errorMsg;
      return false;
    }
    return true;
  }
}
